backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java

```diff
@@ -99,45 +109,86 @@ public LiveOktaRepository(
 
     // By default, when creating a user, we give them privileges of a standard user
     String organizationExternalId = org.getExternalId();
-    Set<OrganizationRole> roles = EnumSet.of(OrganizationRole.getDefault(), role);
-    Set<String> groupIds = new HashSet<>();
-
-    for (OrganizationRole r : roles) {
-      String groupName = generateGroupName(organizationExternalId, r);
-
-      // Okta SDK's way of getting a group by group name
-      GroupList groups = _client.listGroups(groupName, null, null);
-      if (groups.stream().count() == 0) {
-        throw new IllegalGraphqlArgumentException(
-            String.format("Cannot add Okta user to nonexistent group=%s", groupName));
-      }
-      Group group = groups.single();
-      groupIds.add(group.getId());
+    Set<OrganizationRole> rolesToCreate = EnumSet.of(OrganizationRole.getDefault());
+    rolesToCreate.addAll(roles);
+
+    // Add user to new groups
+    Set<String> groupNamesToAdd = new HashSet<>();
+    groupNamesToAdd.addAll(
+        rolesToCreate.stream()
+            .map(r -> generateRoleGroupName(organizationExternalId, r))
+            .collect(Collectors.toSet()));
+    groupNamesToAdd.addAll(
+        facilities.stream()
+            // use an empty set of facilities if user can access all facilities anyway
+            .filter(f -> !PermissionHolder.grantsAllFacilityAccess(rolesToCreate))
+            .map(f -> generateFacilityGroupName(organizationExternalId, f.getInternalId()))
+            .collect(Collectors.toSet()));
+
+    GroupList orgGroups =
+        _client.listGroups(generateGroupOrgPrefix(organizationExternalId), null, null);
+    if (orgGroups.stream().count() == 0) {
+      throw new IllegalGraphqlArgumentException(
+          String.format(
+              "Cannot add Okta user to nonexistent organization=%s", organizationExternalId));
     }
+    Set<String> orgGroupNames =
+        orgGroups.stream().map(g -> g.getProfile().getName()).collect(Collectors.toSet());
+    groupNamesToAdd.stream()
+        .filter(n -> !orgGroupNames.contains(n))
+        .forEach(
+            n -> {
+              throw new IllegalGraphqlArgumentException(
+                  String.format("Cannot add Okta user to nonexistent group=%s", n));
+            });
+    Set<String> groupIdsToAdd =
+        orgGroups.stream()
+            .filter(g -> groupNamesToAdd.contains(g.getProfile().getName()))
+            .map(g -> g.getId())
+            .collect(Collectors.toSet());
 
     UserBuilder.instance()
         .setProfileProperties(userProfileMap)
-        .setGroups(groupIds)
+        .setGroups(groupIdsToAdd)
         .buildAndCreate(_client);
 
-    return Optional.of(new OrganizationRoleClaims(organizationExternalId, roles));
+    List<OrganizationRoleClaims> claims = _extractor.convertClaims(groupNamesToAdd);
+    if (claims.size() != 1) {
```

Comment:
This seems like an unreachable branch, unless we add in the existing groups from the profile (which we probably shouldn't).